### PR TITLE
RFC 11: Change IPR to use ILP Packet and PSK

### PR DESCRIPTION
**Pull request by emschwartz**
depends on https://github.com/interledger/rfcs/pull/166, https://github.com/interledger/rfcs/pull/175
* change the IPR format to only include a packet and condition
* replace JSON format with binary
* recommend using a PSK implementation to generate the payment request

resolves https://github.com/interledger/rfcs/issues/138

### CODE REVIEW

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 1]
```diff
@@ -26,70 +28,31 @@ This document does **not** specify how payment requests are communicated from a
...
-7. Recipient checks the transfer amount against the ILP Packet and checks the payment request has not expired.
-8. Recipient regenerates the condition and fulfillment from the ILP Packet attached to the incoming transfer, using the same algorithm as before, and fulfills the transfer's condition. If the condition generated does not match the one in the transfer it means that the ILP Packet has been tampered with and the transfer goes unfulfilled.
+1. Recipient creates the ILP Packet and generates the condition using one of algorithms outlined in [Condition Generation](#condition-generation).
+2. Recipient communicates the ILP Packet and condition to the sender.
+3. Sender prepares a local ledger transfer to a connector, held pending the execution condition provided by the recipient and with the ILP Packet provided by the recipient attached.
+4. Connectors forward the payment until it reaches the recipient.
+5. Recipient receives a notification of a held transfer with an ILP Packet attached.
+6. Recipient checks the transfer amount against the ILP Packet and checks the payment request has not expired.
```
Comment (sharafian) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: If the data format that the receiver uses is left undefined, it seems strange to describe the behavior in steps 6 and 7.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 2]
```diff
@@ -2,21 +2,23 @@
...
+
+Receiver-generated conditions are also useful in building non-repudiable application layer protocols. The recipient would generate the condition and cryptographically sign a statement that the preimage of the specified hash indicates the sender has settled a particular debt. The recipient would only disclose the hash preimage upon receipt of payment. Once the sender gets the preimage, they would be able to demonstrate to 3rd parties that they paid the recipient using the preimage and the original signed statement.
 
-A common approach is for conditions to be generated and fulfilled by the receivers of Interledger transfers. This document specifies one such approach that enables receivers to verify that an incoming transfer matches a payment request they previously created, without the receiver needing to keep a record of all outstanding payment requests.
+In cases where non-repudiability is unnecessary, the [Pre-Shared Key](../0016-pre-shared-key/0016-pre-shared-key.md) transport protocol may be preferrable because it does not require end-to-end communication for each payment.
```
Comment (sharafian) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: This should give a clearer definition of what cases you want IPR, and should recommend that PSK should be used unless you really know what you're doing.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md (hunk 1 above)
Comment (emschwartz) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: I'd put that in the same category as including the way the receiver generates the shared secret in PSK. It's not technically required but it's a _very good_ idea and should be followed unless you really know what you're doing.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 3]
```diff
@@ -99,42 +62,17 @@ Additional binary headers, encoded as a base64 string, that should be included i
...
+Recipients MUST include a payment request-specific nonce in the ILP packet to ensure that conditions are unique per-request.
 
-This is how a [PREIMAGE-SHA-256 Crypto Condition](../0002-crypto-conditions) MAY be generated using an HMAC of a JSON-encoded Interledger Payment Request:
+Recipients SHOULD include an expiry date so that they do not accept incoming payments that are paid too long after their creation.
 
-1. The recipient takes the `address`, `amount`, `expires_at`, and `data` fields from the payment request. Any fields that are used for routing are probably not important to perform an integrity check upon, because they may be modified in transit and there is no further use for them once the ILP packet is received attached to an incoming transfer.
-2. The recipient uses a canonical encoding, such as [Canonical JSON](https://www.npmjs.com/package/canonical-json) to create a stringified version of `1.`
-3. The recipient uses a secret key to create an HMAC of `2.`
-4. The recipient uses the digest of `3.` as the preimage of a [PREIMAGE-SHA-256 Crypto Condition](../0002-crypto-conditions).
+### Using a PSK Implementation
```
Comment (justmoon) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: What do these three headlines mean? It sounds like I can use either JSON, Binary or a PSK Implementation, but then the actual text doesn't fit that at all.

If the packet is generated by PSK and you need the packet to get the condition for the payment request, I'd probably reorder these sections.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 4]
```diff
@@ -99,42 +62,17 @@ Additional binary headers, encoded as a base64 string, that should be included i
...
 
-The following snippet shows how this can be done in JavaScript:
+It is RECOMMENDED that recipients use an implementation of the [Pre-Shared Key](../0016-pre-shared-key/0016-pre-shared-key.md) protocol to generate the packet and condition for IPR.
```
Comment (justmoon) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: If I don't use PSK, what would I do instead? How do I decide? Why is PSK recommended?

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 5]
```diff
@@ -99,42 +62,17 @@ Additional binary headers, encoded as a base64 string, that should be included i
...
 
-```js
-const crypto = require('crypto') // Node.js crypto
-const stringify = require('canonical-json') // https://www.npmjs.com/package/canonical-json
-const cc = require('five-bells-condition') // https://www.npmjs.com/package/five-bells-condition
+The recipient should follow BOTH the steps normally followed by the recipient AND sender. The recipient should generate the secret in the normal fashion but MUST NOT share it. The recipient then follows the steps the sender would do in a PSK payment to create the packet and condition from the secret and the payment parameters.
```
Comment (justmoon) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: > The recipient should follow BOTH the steps normally followed by the recipient AND sender.

This sentence kinda makes it sound more complicated than it actually is, but I'm not sure how I would reword it.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 6]
```diff
@@ -99,42 +62,17 @@ Additional binary headers, encoded as a base64 string, that should be included i
...
-    "re": "dinner the other night"
-  }
-})
+For details, see the [Pre-Shared Key Pseudocode](../0016-pre-shared-key/0016-pre-shared-key.md#pseudocode).
 
-const hmac = crypto.createHmac('sha256', RECIPIENT_HMAC_KEY)
-hmac.update(requestString)
-const cryptoCondition = new cc.PreimageSha256()
-cryptoCondition.setPreimage(hmac.digest())
+Note the recipient's use of a PSK implementation does not matter to the sender, because the sender treats the `data` in the ILP packet as opaque and uses the condition from the recipient without needing to know how the fulfillment is generated.
```
Comment (justmoon) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: I think this paragraph should follow right after the statement that PSK is "RECOMMENDED", because it gives some important background as to why it isn't *required*.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 7]
```diff
@@ -99,42 +62,17 @@ Additional binary headers, encoded as a base64 string, that should be included i
 
 It is RECOMMENDED that recipients use a Message Authentication Code (MAC) of the details of their payment request in the condition so that the condition can be used as an integrity check on incoming transfers. This allows a recipient to ensure that a transfer matches a request they previously generated, without the recipient needing to maintain a log of all outstanding payment requests.
 
-Recipients MUST include a payment request-specific component in the account ILP Address to ensure that conditions are unique per-request.
-
-### Preimage Condition from JSON
+Recipients MUST include a payment request-specific nonce in the ILP packet to ensure that conditions are unique per-request.
```
Comment (justmoon) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: This sounds like something I have to worry about in addition to using PSK, but doesn't using PSK take care of this?

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 8]
```diff
@@ -99,42 +62,17 @@ Additional binary headers, encoded as a base64 string, that should be included i
 
 It is RECOMMENDED that recipients use a Message Authentication Code (MAC) of the details of their payment request in the condition so that the condition can be used as an integrity check on incoming transfers. This allows a recipient to ensure that a transfer matches a request they previously generated, without the recipient needing to maintain a log of all outstanding payment requests.
```
Comment (justmoon) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: It's not so much that you're adding a MAC, it's that you're using the condition (which you need to pick anyway) *as* a MAC. So I think something like "you should use the condition as a mac by making the fulfillment a hash of the packet" would be easier to understand.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 9]
```diff
@@ -2,21 +2,23 @@
...
 
-A common approach is for conditions to be generated and fulfilled by the receivers of Interledger transfers. This document specifies one such approach that enables receivers to verify that an incoming transfer matches a payment request they previously created, without the receiver needing to keep a record of all outstanding payment requests.
+In most cases, non-repudiability is unnecessary, as it is sufficient for the sender to get proof that the recipient received the payment even if that proof cannot be used to convince 3rd parties. In such cases, the [Pre-Shared Key](../0016-pre-shared-key/0016-pre-shared-key.md) transport protocol is recommended because it does not require end-to-end communication for each payment.
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: "for each payment"→"for every payment"

Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: "3rd parties"→"third parties"

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 10]
```diff
@@ -26,70 +28,31 @@ This document does **not** specify how payment requests are communicated from a
...
 
-1. Recipient creates the payment request and generates the condition using one of algorithms outlined in [Condition Generation](#condition-generation).
-2. Recipient communicates the payment request to the sender.
-3. Sender generates an ILP Packet from the payment request.
-4. Sender prepares a local ledger transfer to a connector, held pending the execution condition given in the payment request and with the ILP Packet attached.
-5. Connectors forward the payment until it reaches the recipient.
-6. Recipient receives a notification of a held transfer with an ILP Packet attached.
-7. Recipient checks the transfer amount against the ILP Packet and checks the payment request has not expired.
-8. Recipient regenerates the condition and fulfillment from the ILP Packet attached to the incoming transfer, using the same algorithm as before, and fulfills the transfer's condition. If the condition generated does not match the one in the transfer it means that the ILP Packet has been tampered with and the transfer goes unfulfilled.
+1. Recipient creates the ILP Packet and generates the condition using one of algorithms outlined in [Condition Generation](#condition-generation).
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: Let's make this more proper English. "_The_ recipient creates..." (and same on the following steps)

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 11]
```diff
@@ -26,70 +28,31 @@ This document does **not** specify how payment requests are communicated from a
...
+2. Recipient communicates the ILP Packet and condition to the sender.
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: Am I mistaken, or is this data structure containing the "ILP Packet and condition" called the "IPR"?

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 12]
```diff
@@ -26,70 +28,31 @@ This document does **not** specify how payment requests are communicated from a
...
+3. Sender prepares a local ledger transfer to a connector, held pending the execution condition provided by the recipient and with the ILP Packet provided by the recipient attached.
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: Would be great to link to an explanation of "connectors" in the Architecture RFC or somewhere. The term doesn't make much sense if you're not familiar with Interledger.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md (hunk 4 above)
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: Maybe it's better, instead of recommending PSK, to define and recommend a system that is very, very similar to PSK.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 13]
```diff
@@ -2,21 +2,23 @@
 
 ## Preface
 
-This document specifies the Interledger Payment Request (IPR) data format. It is intended for use in protocols in which the recipient of an Interledger payment first communicates a request for payment to the sender.
-
-This document provides an example flow to illustrate how IPR could be used in a higher level protocol. This document also details a method for generating Interledger transfer conditions that enable recipients to use the condition as an integrity check on incoming transfers and packets, without maintaining state about all outstanding payment requests.
+This document specifies the Interledger Payment Request (IPR) transport protocol. It is an end-to-end protocol in which the recipient of an Interledger payment first communicates a request for payment to the sender. This document also details how the recipient can verify that incoming payments match previously created requests without storing a history of all outstanding requests.
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: > ## Preface
> This document specifies the Interledger Payment Request (IPR) transport protocol. It is an end-to-end protocol...

I suggest you drop the "Preface" header and change the opening sentences to read:

> The Interledger Payment Request (IPR) transport protocol is an end-to-end protocol...

File: 0011-interledger-payment-request/0011-interledger-payment-request.md (hunk 4 above)
Comment (sharafian) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: The reason that the IPR spec recommends PSK is that in actual implementation, IPR and PSK are 99% identical. The only difference is that the `generatePskParams` is called per payment in IPR, and at protocol setup for PSK.

**Comment by emschwartz** [comment]
@mDuo13 I agree with @sharafian that it doesn't make sense to redefine the data format and steps since they would mostly duplicate what is written in the PSK spec. I tried to be clearer about why PSK is being used and how the receiver should actually do it. Let me know if you still think we should copy over the steps and pseudocode from PSK.

@justmoon I reordered the sections, let me know if you think that works better.

Also, since the JSON format now has only 2 fields and both of them are binary, I think we might want to get rid of the JSON representation and just make a binary format that puts the condition first and the packet right after it. @justmoon do you want to make an ASN.1 definition or should we just say you concatenate them (you'd know that the condition is the first 32 bytes)?

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 14]
```diff
@@ -1,140 +1,85 @@
...
+### Using a PSK Implementation
 
-#### amount
+The [Pre-Shared Key](../0016-pre-shared-key/0016-pre-shared-key.md) protocol uses the payment condition as a MAC of the payment details by setting the fulfillment to be an HMAC of the ILP packet with a key derived from a secret shared between the sender and receiver. In IPR there is no shared secret and the condition is only generated by the receiver. Nevertheless, an IPR receiver can use the same methods to generate the condition and regenerate the fulfillment on receipt of the incoming transfer.
 
-The amount requested by the recipient, such as `10.25`, denoted in the assets of the recipient's ledger.
+To use a PSK implementation, the receiver follows these steps:
 
-The amount MUST be rounded based on the precision supported by the recipient's ledger.
+1. The receiver MAY derive a receiver secret and address as detailed in [PSK: Shared Secret Generation](../0016-pre-shared-key/0016-pre-shared-key.md#shared-secret-generation) but the receiver DOES NOT share the secret key.
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: "DOES NOT"→"MUST NOT"

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 15]
```diff
@@ -1,140 +1,85 @@
...
+2. The receiver creates the payment packet and condition _using the steps normally executed by the sender_: [PSK: Payment Creation](../0016-pre-shared-key/0016-pre-shared-key.md#1-payment-creation).
+3. The receiver communicates the payment request, including the packet and condition, to the sender.
+4. If the receiver used the derived secret from `1.`, they regenerate that secret from the packet as specified in [PSK: Shared Secret Regeneration](../0016-pre-shared-key/0016-pre-shared-key.md#shared-secret-regeneration).
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: Instead of writing `1.` in code brackets, just say "step 1" here.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 16]
```diff
@@ -1,140 +1,85 @@
...
+5. The receiver generates the fulfillment from the secret and packet as specified in [PSK: Payment Fulfillment](../0016-pre-shared-key/0016-pre-shared-key.md#2-payment-fulfillment).
 
-Recipients SHOULD ignore incoming transfers whose amounts are less than the amount specified in the payment request. Recipients MAY ignore incoming transfers whose amounts are greater than requested.
+### Alternative Methods and General Best Practices
 
-#### condition
+Receivers MAY use any method they choose for generating the condition, such as using a random fulfillment for each request. However, this also requires storing every outstanding fulfillment and packet in a database.
 
-The Crypto Condition to be used as the execution condition for the payment. See [Condition Generation](#condition-generation) for recommendations the recipient MAY use for creating the condition.
+A PSK implementation will already provide these features, but receivers that choose a different method for generating the conditions SHOULD consider these recommendations:
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: It's not clear what "these features" means here—it almost reads like a PSK implementation includes storing every outstanding fulfillment in a database. Also, avoid future tense ("will provide") unless you're talking about actual future events.

I suggest rephrasing this section as follows:

> Receivers MAY use any method they choose for generating the condition, such as using a random fulfillment for each request. The receiver MUST be able to identify and verify the ILP Packet for every outstanding payment and produce the corresponding fulfillment. A PSK implementation handles these requirements without storing every fulfillment and packet in a database, by using the condition as a Message Authentication Code (MAC) of the payment request.
> 
> Receivers SHOULD follow these guidelines if implementing a custom system:
>
> - Recivers SHOULD include a nonce in the ILP Packet to distinguish it from other packets that would otherwise be identical. Each payment request SHOULD have a unique nonce.
> - Receivers SHOULD include an expiry date in the ILP Packet, so they do not accept incoming packets that are paid much later than their creation date.

File: 0011-interledger-payment-request/0011-interledger-payment-request.md [hunk 17]
```diff
@@ -1,140 +1,85 @@
...
+|:--|:--|:--|
+| `packet` | [Base64-Url](https://en.wikipedia.org/wiki/Base64#URL_applications) Encoded ILP Packet | ILP Payment packet including the destination address and amount |
+| `condition` | 32 Bytes, Base64-Url Encoded | Execution condition for the payment |
 
-Additional binary headers, encoded as a base64 string, that should be included in the ILP Packet. The sender SHOULD treat these as opaque.
+In IPR the sender only uses the `account` and `amount` from the ILP packet and MUST treat the `data` as opaque. The sender MUST NOT modify the ILP packet at all, or the receiver will reject the payment.
 
 ### Binary
 
 **TODO**
```
Comment (mDuo13) [inline 0011-interledger-payment-request/0011-interledger-payment-request.md]: Don't merge this request while this **TODO** is still here.

**Comment by emschwartz** [comment]
@mDuo13's comments have been addressed so I'm going to go ahead and merge this
